### PR TITLE
fix: Transform `ObjectExistsWithSize` into `GetAttributes`

### PR DESCRIPTION
**Pull request by DylanGuedes**
**What this PR does / why we need it**:
- Transform our `ObjectExistsWithSize` API into `GetAttributes` to conform with the existing Thanos API.
- Fix we not returning the size when we should on our s3 implementation of `ObjectSize`
- Fix we not returning `false, nil` to `ObjectExists` when the only error seen is a `NotFound`. If someone calls `ObjectExists` it should return `false, nil` to a `NotExist` error, instead of `false, err`

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] Title matches the required conventional commits format, see [here](https://www.conventionalcommits.org/en/v1.0.0/)
  - **Note** that Promtail is considered to be feature complete, and future development for logs collection will be in [Grafana Alloy](https://github.com/grafana/alloy). As such, `feat` PRs are unlikely to be accepted unless a case can be made for the feature actually being a bug fix to existing behavior.
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/storage/chunk/client/azure/blob_storage_client.go [hunk 1]
```diff
@@ -220,13 +220,17 @@ func NewBlobStorage(cfg *BlobStorageConfig, metrics BlobStorageMetrics, hedgingC
...
+	_, err := b.objectAttributes(ctx, objectKey, "azure.ObjectExists")
+	return err == nil, err
 }
 
-func (b *BlobStorage) ObjectExistsWithSize(ctx context.Context, objectKey string) (bool, int64, error) {
+func (b *BlobStorage) GetAttributes(ctx context.Context, objectKey string) (client.ObjectAttributes, error) {
+	return b.objectAttributes(ctx, objectKey, "azure.GetAttributes")
+}
+
+func (b *BlobStorage) objectAttributes(ctx context.Context, objectKey, source string) (client.ObjectAttributes, error) {
```
Comment (JoaoBraveCoding) [inline pkg/storage/chunk/client/azure/blob_storage_client.go:231]: nit: IIUC using a different `source` is just going to create a different metric? Does it make sense to separate calls to this function into separate metrics? 
My suggestion would be to just drop the `source` parameter, if I'm understanding the code correctly

**Review by JoaoBraveCoding — COMMENTED** [review]
Small nits, overall lgtm! Thank you for this!

File: pkg/storage/chunk/client/azure/blob_storage_client.go (hunk 1 above)
Comment (DylanGuedes) [reply pkg/storage/chunk/client/azure/blob_storage_client.go:236]: Yes, the idea is to allow disambiguation between `GetAttributes` and `ObjectExists` where it is possible to do it. Loki isn't using `GetAttributes` so this shouldn't cause any problems but if you're using Loki's API for doing something, having this separation will help your application separating what is leading the object storage requests.

File: pkg/storage/chunk/client/gcp/gcs_object_client.go [hunk 2]
```diff
@@ -127,20 +127,23 @@ func (s *GCSObjectClient) Stop() {
 }
 
 func (s *GCSObjectClient) ObjectExists(ctx context.Context, objectKey string) (bool, error) {
-	exists, _, err := s.ObjectExistsWithSize(ctx, objectKey)
-	return exists, err
+	_, err := s.GetAttributes(ctx, objectKey)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
```
[the remark this answers is not in the document]
Comment (DylanGuedes) [reply pkg/storage/chunk/client/gcp/gcs_object_client.go]: That's super fair but I found a problem with the approach I was following so I had to rewrite everything. I should only return an error if the error isn't `NotFound`. Not all clients support it which is ok but some of them support it so the rewrite was necessary to keep the old behavior.

**Review by paul1r — APPROVED** [review]

**Pull request merged**